py3status/modules/yubikey.py

```diff
@@ -0,0 +1,183 @@
+                            # if this doesn't return very quickly, touch is pending
+                            check.communicate(
+                                timeout=self.parent.gpg_check_timeout)
+                            time.sleep(0.25)
+                        except subprocess.TimeoutExpired:
+                            self._set_pending(True)
+                            os.killpg(check.pid, signal.SIGINT)
+                            check.communicate()
+                            # wait until device is touched
+                            with self._check_card_status() as wait:
+                                wait.communicate()
+                                self._set_pending(False)
+                                break
+
+                # ignore all events caused by operations above
+                for _ in inotify.read():
+                    pass
+
+        for watch in watches:
+            inotify.rm_watch(watch)
+
+
+class U2fWatchingThread(threading.Thread):
+    """
+    A thread watchng if YubiKey is waiting for a touch for u2f access
+    """
+
+    def __init__(self, parent):
+        super(U2fWatchingThread, self).__init__()
+        self.parent = parent
+
+    def _set_pending(self, new_value):
+        self.parent.pending_u2f = new_value
+        if not self.parent.pending_gpg:
+            self.parent.py3.update()
+
+    def run(self):
+        inotify = inotify_simple.INotify()
+        watches = []
+        for path in self.parent.u2f_check_watch_paths:
+            if os.path.isfile(path):
+                watch = inotify.add_watch(path, inotify_simple.flags.OPEN)
+                watches.append(watch)
+
+        while not self.parent.killed.is_set():
+            for event in inotify.read():
+                self._set_pending(True)
+                for event in inotify.read():
+                    self._set_pending(False)
+
+        for watch in watches:
+            inotify.rm_watch(watch)
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    format = '\?if=is_waiting YubiKey'
+    gpg_check_timeout = 0.1
+    gpg_check_watch_paths = ['~/.gnupg/pubring.kbx', '~/.ssh/known_hosts']
+    u2f_check_watch_paths = ['~/.config/Yubico/u2f_keys']
+
+    def post_config_hook(self):
+        self.gpg_check_watch_paths = [
+            os.path.expanduser(p) for p in self.gpg_check_watch_paths
+        ]
+        self.u2f_check_watch_paths = [
+            os.path.expanduser(p) for p in self.u2f_check_watch_paths
+        ]
+
+        self.killed = threading.Event()
+
+        self.pending_gpg = False
+        self.pending_u2f = False
+
+        if len(self.gpg_check_watch_paths) > 0:
+            GpgWatchingThread(self).start()
+
+        if len(self.u2f_check_watch_paths) > 0:
+            U2fWatchingThread(self).start()
+
+    def yubikey(self):
+        is_waiting = self.pending_gpg or self.pending_u2f
+        format_params = {'is_waiting': is_waiting}
```

Comment:
Do more with formatting tricks by replacing `is_waiting` with `gpg` and `u2f`, eg...
```
format = '[\?color=#fa0 YubiKey [\?if=gpg&color=#aaf GPG][\?if=u2f&color=#faf U2F]]'
format = '[\?if=gpg&color=degraded YubiKey GPG][\?if=u2f&color=#fa0 YubiKey U2F]]'
format = '[\?color=good YubiKey [\?if=gpg GPG][\?if=u2f U2F]]'
format = '[\?color=good [\?if=gpg GPG][\?if=u2f U2F]]'
format = '[\?color=good&u2f U2F]'  # ignore GPG
```

Reply:
This is interesting thought, but in practice this would be just too much. Users don't care whether the module is flashing because of `gpg` or `u2f`, once it flashes for whatever reason they need to figure out manually what caused this and decide on the spot, touch or not touch.

What could have been interesting is to see the process name that requests the touch, e.g. I can configure using both `gpg` or `u2f` technology that `ssh`'ing somewhere will ask me to touch the device for confirmation, and in real life I don't care which technology is being used, instead I want to know which process wants this confirmation. But I don't know how to detect that from the module, we will leave this to a different time 🙂

Reply:
>Users don't care whether the module is flashing because of gpg or u2f, once it flashes for whatever reason they need to figure out manually what caused this and decide on the spot, touch or not touch.

Possibly all the more reasons to do... as GPG is related to security, keys, encryption, logins, credentials, etc. Some users may like that option or may use nice icons to differentiate.

>once it flashes for whatever reason they need to figure out manually what caused this

It at least differentiates `Something to do with GPG`. Hopefully it will grow more from there... to support `{process_name}`, `{pid}`, Idk what else.

>maximbaz (in another thread): Of course it would be awesome if one of the yubikey CLI tools provided such information out of the box, it is a really useful feature in my mind...

If that happen, then we should use it to expose everything for users... just like `cmus`, `moc` and `coin_market` as they all have placeholders the users may not care about. We included them all because the data is there.

Reply:
Let's first find at least one person who needs that 🙂